### PR TITLE
[4.0] Group toolbar buttons [com-content -> articles]

### PR DESCRIPTION
**Pull request by ciar4n**
Pull Request for Issue https://github.com/joomla/40-backend-template/issues/348 .

### Summary of Changes
Groups com_content -> articles toolbar buttons

Credit to @asika32764 for adding #19670

### Testing Instructions
Apply PR and navigate to com_content -> articles.

### Before PR
image

### After PR
image

image

### Documentation Changes Required
Yes

### CODE REVIEW

**Review by brianteeman — APPROVED** [review]
approve new language string

**Comment by brianteeman** [comment]
This loses the work that @fedik did to disable buttons

**Comment by ciar4n** [comment]
@brianteeman Would you have the number on that PR?

**Comment by ciar4n** [comment]
Any suggestions for the 'Change Status' icon and color?

**Comment by brianteeman** [comment]
I think the pr was #20650 

You will need to get the a11y team to review this so that the dropdown is described and announced correctly

**Comment by ciar4n** [comment]
> You will need to get the a11y team to review this so that the dropdown is described and announced correctly

@joomla/joomla-accessibility-team-jat Please review.

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 47fc7c140089f214d5960b0945dec86feaa3afaa<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21247">issues.joomla.org/tracker/joomla-cms/21247</a>.</sub>

**Comment by ciar4n** [comment]
> This loses the work that @Fedik did to disable buttons

@Fedik As JS is not really my forte, I'll ask you for some guidance on this.

**Comment by chmst** [comment]
I would like to remember my PR and show only buttons if they can do someting . https://issues.joomla.org/tracker/joomla-cms/20559

**Comment by dgrammatiko** [comment]
@ciar4n are you using Bootstrap's dropdown here?

**Comment by ghost** [comment]
@chmst 
> show only buttons if they can do someting

+1

**Comment by ciar4n** [comment]
> @ciar4n are you using Bootstrap's dropdown here?

Yes. I think doing anything else is outside of the scope of this PR. For now lets consider that grouping buttons is the way to go and if so, how best to group them,

**Comment by brianteeman** [comment]
no - lets wait a year for a custom element - that solves everything everywhere

**Comment by dgrammatiko** [comment]
@brianteeman you know what I'll revert all the custom elements work. Back to jQuery bootstrap seems what people want! So let me please you and good luck with Joomla 4...

**Comment by wilsonge** [comment]
Please stop fighting. Ciaran is correct that this is not the place for the custom elements discussion - he's using the api for grouping buttons. Whether that uses Bootstrap or Custom Elements is a totally different issue/PR. (To an extent so is the accessibility of that dropdown as it's not just going to be used here - but across many extensions for differing groupings - although of course if we can sort that here all the better!).

However ensuring that we don't regress the grey'ing out of buttons when no items are selected is important and should be considered as part of this issue

**Comment by Fedik** [comment]
the question here, what should be "grayed" (disabled), the "main" button of the group, or the elements in it?

the elements not grayed, because in the group they are "anchors" `<a>` not `<button>`, if I right remember,
I try to check more next weekend

**Comment by brianteeman** [comment]
@Fedik As none of the "change status" options will be enabled unless an item is selected then I would suggest that it should be the main button that is disabled

**Comment by chmst** [comment]
My2Cent: The graying needs a designer. As it is now, I always try to clean my glasses. Maybe gray could be really gray? <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21247">issues.joomla.org/tracker/joomla-cms/21247</a>.</sub>

**Comment by brianteeman** [comment]
Or even grey 😂

**Comment by ghazal** [comment]
I have tested this item :white_check_mark: successfully on 47fc7c140089f214d5960b0945dec86feaa3afaa<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21247">issues.joomla.org/tracker/joomla-cms/21247</a>.</sub>

**Comment by ghost** [comment]
Ready to Commit after two successful tests.

**Comment by wilsonge** [comment]
Please remove the RTC as per discussions above here https://github.com/joomla/joomla-cms/pull/21247#issuecomment-407915376 it is not ready to be merged yet

**Comment by zwiastunsw** [comment]
There was a request for an accessibility test. The toolbar does not meet the accessibility requirements. I am preparing a report. But I need time

**Comment by brianteeman** [comment]
@zwiastunsw for future reference if you are doing an a11y test please can you post that it is in progress so that the pr doesnt get merged while you are testing

**Comment by zwiastunsw** [comment]
Of course, the last days I were very busy.

**Comment by brianteeman** [comment]
@zwiastunsw thats not a problem - just so that we know its being done

**Comment by wojsmol** [comment]
@franz-wohlkoenig Please remove RTC label on tracker, see https://github.com/joomla/joomla-cms/pull/21247#issuecomment-408391417 and below.

**Comment by ghost** [comment]
@wojsmol done, Status is now "Pending".

**Comment by Fedik** [comment]
I can make "disabled" the elements in dropdown, 
but I hit the problem, with the main button, 
Bootstrap crashes if I add `<joomla-toolbar-button/>` to toggler:

```html
<joomla-toolbar-button>
 <button class="button-group btn btn-info dropdown-toggle" 
    data-toggle="dropdown">Bla</button>
</joomla-toolbar-button>
```
Bootstrap do not expect extra wrapper...

@dgrammatiko maybe you have some idea?

**Comment by Fedik** [comment]
@ciar4n please check https://github.com/ciar4n/joomla-cms/pull/4

**Comment by dgrammatiko** [comment]
@Fedik use this: https://github.com/joomla-projects/custom-elements/blob/master/src/js/dropdown/dropdown.js

**Comment by Fedik** [comment]
~~hm, then I have no idea :)~~

**Comment by Fedik** [comment]
okay, fixed,
@ciar4n please check https://github.com/ciar4n/joomla-cms/pull/4 
this should work now

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 47fc7c140089f214d5960b0945dec86feaa3afaa<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21247">issues.joomla.org/tracker/joomla-cms/21247</a>.</sub>

**Comment by Fedik** [comment]
@franz-wohlkoenig wait, @ciar4n still need to merge my pull request :)

**Comment by ciar4n** [comment]
Thank you @Fedik 👍

**Comment by zwiastunsw** [comment]
If the purpose of this PR is only to group buttons, I have no comments. However, I have to write here that **this toolbar causes accessibility problems**. I have presented them here: [Toolbar - About accessibility](https://github.com/joomla/accessibility/issues/42). 

Concerning the "disabled" of the items in the dropdown: the intention is right, but in this case it should be the buttons and not the anchors. The "Publish", "Unpublish", "Feature", "Unfeature", "Archive", "Trash" options do not move the user to another page or place, but only change the status of the selected items.

**Comment by chmst** [comment]
I think there are two issues. First: grouping the buttons. This  looks good.

The second is: Disable buttons when no item is selected. Do we really need this disabling? 
The message in J3:  "Please make a selection" is quite clear. I never saw a question or complaint there. In the contray. Now i must know what to do for enabling the buttons.

**Comment by dgrammatiko** [comment]
> Do we really need this disabling?

Actually,  from UX perspective, disabling is not the best option. If something is not available you should not present it, that is the rule of thumb for good UX

**Comment by chmst** [comment]
As I said .. but in this case I see this: The function is available as soon as a singel item is in the list and the user has the right to change the status. So the buttons can be used but need checked items. A message "Please check items" is correct and we must do nothing

If the user has not the edit.state right or if the list is empty: https://issues.joomla.org/tracker/joomla-cms/20559
But this can be another PR, we must not mix these two issues

**Comment by zwiastunsw** [comment]
> Actually, from UX perspective, disabling is not the best option. If something is not available you should not present it, that is the rule of thumb for good UX

I am not a UX specialist. This is a toolbar. I should see what kind of toolkit I have. But it is also common (and good?) practice if options that I can't use are disabled. So I do not think that failure to show is a good thing.

**Comment by Fedik** [comment]
> Concerning the "disabled" of the items in the dropdown: the intention is right, but in this case it should be the buttons and not the anchors

then just need to replace `<a>` to `<button>`, and problem solved

**Comment by ciar4n** [comment]
I've change dropdown items from anchors to buttons and grey'd disabled items as suggested.

File: administrator/templates/atum/scss/blocks/_toolbar.scss [hunk 1]
```diff
@@ -49,23 +49,21 @@
       color: #fff;
       background-color: rgba(0, 0, 0, .2);
     }
+
+    &[disabled='true'],
```
Comment (Fedik) [inline administrator/templates/atum/scss/blocks/_toolbar.scss]: it enough just `&[disabled]` or `&:disabled`

Comment (ciar4n) [reply administrator/templates/atum/scss/blocks/_toolbar.scss]: True. Amended with last commit.

**Comment by ciar4n** [comment]
PR should be ready for testing if anybody is willing 😄

**Comment by ggppdk** [comment]
I have tested this item :white_check_mark: successfully on d23fd847995aee1511895d182b07ac665cc87408<br /><br />Great improvement
Scanning so many big buttons horizontally is a headache / tiresome 
plus it is intimidating to see them there all the time
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21247">issues.joomla.org/tracker/joomla-cms/21247</a>.</sub>

**Comment by ggppdk** [comment]
I imagine other core screens (and 3rd party extensions), will follow after this is merged

**Comment by ghost** [comment]
Applying Patch got
screen shot 2018-07-31 at 13 08 12
and Patch isn't applied.

**Comment by mbabker** [comment]
That's a patch tester issue, nothing to do with this...

**Comment by ghost** [comment]
thanks Michael, but Patch isn't applied.

**Comment by mbabker** [comment]
By design.  After filtering out files that shouldn't be patched (there are a bunch of files in the repo that don't exist in release packages), if a file that is supposed to be patched doesn't exist then applying the patch fails.

As the notification says, 4.0 support is experimental because the repo is in a constantly changing state.  Patch tester doesn't know about any of the NPM build tools to ignore those in that filtering step mentioned above, so a pull request changing some of those files isn't going to work correctly.

**Comment by wojsmol** [comment]
Additionally this PR have merge conflicts.

**Comment by wilsonge** [comment]
Ok @ciar4n would you mind sorting out the conflicts here (sorry) if we have the disabled status sorted then I'll get this merged as a working implementation of the grouped buttons and we can then additionally work on ally/custom elements etc :)

**Comment by ciar4n** [comment]
Unable to compile SCSS with the recent media changes. Already spent half the morn with Dimitris trying to sort it out with no luck. Issue seems to be unique to windows.

**Comment by brianteeman** [comment]
Ah. So it wasn't just me then.

**Comment by mbabker** [comment]
If there is one thing I've learned over the last year and a half it's that NPM and Node are probably the worst as it relates to being problematic to use on Windows, everyone at my office has had issues galore since starting new projects or migrating legacy projects that use SCSS compiling (funny enough the LESS compilers work fine, and it all works fine on our Linux dev tools or CI pipelines).

**Comment by brianteeman** [comment]
is this the error you are getting

```
something exploded Invalid CSS after "...eed to override": expected 1 selector or at-rule, was "text-rendering: aut"
something exploded 1
something exploded 7
something exploded Invalid CSS after "...ly-monospace)}}": expected "}", was ""
something exploded 1
something exploded 7
something exploded Invalid CSS after "...ly-monospace)}}": expected "}", was ""
something exploded 1
something exploded 7
something exploded Invalid CSS after "...ly-monospace)}}": expected "}", was ""
something exploded 1
something exploded 31
something exploded Undefined variable: "$input-btn-padding-y".
something exploded 3
something exploded 36
something exploded argument `$color` of `darken($color, $amount)` must be a color
something exploded 129
something exploded 36
something exploded argument `$color` of `darken($color, $amount)` must be a color
something exploded 129

```

**Comment by dgrammatiko** [comment]
@brianteeman yes that was it iirc

**Comment by wojsmol** [comment]
@ciar4n @brianteeman @dgrammatiko scss compilation is working for me on Windows 8.1 if I run `node build.js --compile-css`
```
node -v
v8.11.2
npm -v
6.2.0
```
@dgrammatiko npm run build:css in not working
```
joomla@4.0.0 build:css W:\var\www\4npmtest
> node build.js --compilecss


  error: unknown option `--compilecss'

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! joomla@4.0.0 build:css: `node build.js --compilecss`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the joomla@4.0.0 build:css script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

**Comment by dgrammatiko** [comment]
@wojsmol thanks, this should fix it: https://github.com/joomla/joomla-cms/pull/21350

**Comment by ggppdk** [comment]
This is failing for me too
`npm run build:css`

When i posted successuful test i used: 
(I have npm 5.6.0)
`node build.js --compile-css`

because i noticed that `npm install` (postinstall part) is using the above
will test #21350

**Comment by ciar4n** [comment]
@wilsonge conflicts fixed

**Comment by ciar4n** [comment]
@wilsonge if you can merge this, i can look at doing similar in other views

File: templates/system/incompatible.html [hunk 2]
```diff
@@ -6,7 +6,7 @@
...
+  <script>window.errorLocale = {"c:\xampp\htdocs\joomla-4-dev\installation\language\en-GB\en-GB":{"language":"English GB","header":"Sorry, your PHP version is not supported","text1":"Your host needs to use PHP version {{phpversion}} or newer to run this version of Joomla","help-url-text":"Help me resolve this"},"c:\xampp\htdocs\joomla-4-dev\installation\language\en-US\en-US":{"language":"English US","header":"Sorry, your PHP version is not supported","text1":"Your host needs to use PHP version {{phpversion}} or newer to run this version of Joomla","help-url-text":"Help me resolve this"},}</script><!-- Sets the content of the translated text, IMPORTANT DO NOT REMOVE -->
```
Comment (wilsonge) [inline templates/system/incompatible.html]: @dgrammatiko this looks like a windows build issue if i ever saw one :D

File: libraries/src/MVC/Model/AdminModel.php [hunk 3]
```diff
@@ -928,7 +928,7 @@ public function getItem($pk = null)
 
 		// Convert to the CMSObject before adding other data.
 		$properties = $table->getProperties(1);
-		$item = ArrayHelper::toObject($properties, CMSObject::class);
+		$item = ArrayHelper::toObject($properties, 'CMSObject');
```
Comment (wilsonge) [inline libraries/src/MVC/Model/AdminModel.php]: Can we leave this one as it was before please

**Comment by wilsonge** [comment]
@rdeutz is the system test here failing because we're moving the basic toolbar to a webcomponent or some other reason?

**Comment by wilsonge** [comment]
Definitely want to get this merged. Need Dimitris to fix the installation stuff (you can manually patch you versions for now in both installation files), fix the merge conflict you accidentally committed and I've asked Robert to look into the system test failure because that's not the 'normal' installation one that fails from time to time

**Comment by dgrammatiko** [comment]
> fix the installation stuff

What's broken?

File: templates/system/incompatible.html (hunk 2 above)
Comment (dgrammatiko) [reply templates/system/incompatible.html]: @ciar4n can you test if https://github.com/joomla/joomla-cms/compare/4.0-dev...dgrammatiko:es6/4.0-dev/fix-build-tools?expand=1 fixes this dos problem?

**Comment by infograf768** [comment]
Question: is this still compatible with the WORKFLOW?
(Just compared the test instructions screenshot with what we have now with workflow.)

Please post a screenshot of what is now expected from this PR, specially the Status dropdown.

**Comment by infograf768** [comment]
Also, can't apply here (with eclipse) the changes brought to cassiopea

**Comment by infograf768** [comment]
After patching I confirm that using the "Change Status" toolbar button is useless except for Feature/Unfeature.

**Comment by ciar4n** [comment]
> Question: is this still compatible with the WORKFLOW?

Honestly no idea. All I get when opening 'workflow' is... 

image

I'd have presumed it was a different component so doesn't apply here. I guess that must be incorrect. Also add that this PR was created prior to workflow been merged.

> Also, can't apply here (with eclipse) the changes brought to cassiopea

What changes exactly?

> After patching I confirm that using the "Change Status" toolbar button is useless except for Feature/Unfeature.

I assume this is in the 'workflow' component?

**Comment by infograf768** [comment]
> I assume this is in the 'workflow' component?

Nope. It is in the Articles Manager where the new workflow stuff has impact.

Since workflow has been merged, quite a few things have disapeared from the time when you created this PR. Basically, we have, in Articles Manager, a totally different use of the Toolbar.

**Comment by infograf768** [comment]
Here is what we get now without your patch:
screen shot 2018-08-15 at 09 43 49

**Comment by ciar4n** [comment]
> Here is what we get now without your patch:

~~Certainly not what I am seeing on this end. I think I might have to park this until workflows is more stable.~~

**Comment by ciar4n** [comment]
> After patching I confirm that using the "Change Status" toolbar button is useless except for Feature/Unfeature.

With workflows introduced, this is indeed true. I have no idea how to go about fixing this.

**Comment by ciar4n** [comment]
So it appears workflow removes the publish/unpublish/archive/trash buttons from the toolbar? I guess that leaves feature/unfeatured the only items added to the 'Change Status' dropdown?

**Comment by brianteeman** [comment]
and that is just plain wrong and hopefully it will be changed

**Comment by ciar4n** [comment]
This PR contains the required CSS to apply similar changes to other views. So for the sake of moving things along, only the feature/unfeatured options remain in the toolbar (as is currently the case).  They can be added back when ready by does working on workflows.

This PR should be once again ready for testing/merging

**Comment by bembelimen** [comment]
I really like this PR, perhaps we could implement a in some way "intelligent" dropdown button, which offers transitions depending on the selected articles...

The good thing is, with this approach we would get, although we're using the new workflow, exactly this output, when using the unmodified default workflow: https://user-images.githubusercontent.com/2803503/43141243-6266f50c-8f4d-11e8-8d8a-a80a96bf4127.png

**Comment by coolcat-creations** [comment]
If it could look like https://github.com/joomla/joomla-cms/issues/22004 designwise it would be great.

**Comment by ciar4n** [comment]
Closing. Replaced by #22393.

**Pull request closed without merging**